### PR TITLE
Modified code to wait until controller connects to switch

### LINKED ISSUE #3077: [MME] Fix "controller not connected to switch" error when MME restarts.
Can we add a retry here so that MME does not go into crashloop?

### PR DESCRIPTION
**Pull request by rsarwad**
Modified code to wait until controller connects to switch

### CODE REVIEW

File: lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp [hunk 1]
```diff
@@ -101,22 +107,39 @@ void OpenflowController::dispatch_event(const ControllerEvent& ev) {
 void OpenflowController::inject_external_event(
     std::shared_ptr<ExternalEvent> ev, void* (*cb)(std::shared_ptr<void>) ) {
   if (latest_ofconn_ == NULL) {
-#define CONNECTION_WAIT_TIME 5
+#define CONNECTION_EVENT_WAIT_TIME 5
     std::unique_lock<std::mutex> lck(cv_mutex);
-    if (cv.wait_for(lck, std::chrono::seconds(CONNECTION_WAIT_TIME)) ==
+    if (cv.wait_for(lck, std::chrono::seconds(CONNECTION_EVENT_WAIT_TIME)) ==
```
Comment (pshelar) [inline lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp]: Can you call is_controller_connected_to_switch() function here.

File: lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp [hunk 2]
```diff
@@ -101,22 +107,39 @@ void OpenflowController::dispatch_event(const ControllerEvent& ev) {
...
 }
 
+bool OpenflowController::is_controller_connected_to_switch(void) {
+  /* c++ provided conditional variable is added along with wait for 5 minutes in
+   * order to make sure connection is established between Controller and switch
+   * before inserting the OVS rules
+   */
+
+#define CONNECTION_WAIT_TIME 300
+  std::unique_lock<std::mutex> lck(cv_mutex);
```
Comment (pshelar) [inline lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp:130]: there shld be log msg to indicate the process is waiting connection.

**Review by pshelar — CHANGES_REQUESTED** [review]

**Review by pshelar — APPROVED** [review]
Thanks for exploring all options and finishing this issue.

File: lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp [hunk 3]
```diff
@@ -94,12 +107,37 @@ void OpenflowController::dispatch_event(const ControllerEvent& ev) {
...
+      duration);
+  std::unique_lock<std::mutex> lck(cv_mutex);
+  if (cv.wait_for(lck, std::chrono::seconds(duration)) ==
+      std::cv_status::timeout) {
+    OAILOG_CRITICAL(
+        LOG_GTPV1U,
+        "Failed to connect openflow controller to switch, waited for %d "
+        "seconds \n",
+        duration);
+    return -1;
```
Comment (ssanadhya) [inline lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp:138]: Use RETURNError or RETURNok here

File: lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp [hunk 4]
```diff
@@ -94,12 +107,37 @@ void OpenflowController::dispatch_event(const ControllerEvent& ev) {
 void OpenflowController::inject_external_event(
     std::shared_ptr<ExternalEvent> ev, void* (*cb)(std::shared_ptr<void>) ) {
   if (latest_ofconn_ == NULL) {
-    OAILOG_ERROR(
-        LOG_GTPV1U, "Null connection on event type %d", ev->get_type());
-    throw std::runtime_error("Controller not connected to switch:\n");
+#define CONNECTION_EVENT_WAIT_TIME 5
+    if (!is_controller_connected_to_switch(CONNECTION_EVENT_WAIT_TIME)) {
```
Comment (ssanadhya) [inline lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp:111]: This appears as if we are checking for the failure of `is_controller_connected_to_switch`, instead can we do
```suggestion
    if (is_controller_connected_to_switch(CONNECTION_EVENT_WAIT_TIME) == RETURNok) {
```

File: lte/gateway/c/oai/lib/openflow/controller/OpenflowController.h [hunk 5]
```diff
@@ -115,6 +117,7 @@ class OpenflowController : public fluid_base::OFServer {
    */
   void inject_external_event(
       std::shared_ptr<ExternalEvent> ev, void* (*cb)(std::shared_ptr<void>) );
+  bool is_controller_connected_to_switch(int);
```
Comment (ssanadhya) [inline lte/gateway/c/oai/lib/openflow/controller/OpenflowController.h:120]: Add comment describing the function intent

File: lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp [hunk 6]
```diff
@@ -59,9 +67,15 @@ void OpenflowController::message_callback(
...
-    OAILOG_DEBUG(LOG_GTPV1U, "Openflow controller connected to switch\n");
+    OAILOG_INFO(LOG_GTPV1U, "Openflow controller connected to switch \n");
     // Save OF connection for external events
     latest_ofconn_ = ofconn;
+    std::lock_guard<std::mutex> lck(cv_mutex);
+    cv.notify_all();
+    OAILOG_INFO(
+        LOG_GTPV1U,
+        "Send signal that Controller is connected to switch to all waiting "
+        "threads \n");
```
Comment (ssanadhya) [inline lte/gateway/c/oai/lib/openflow/controller/OpenflowController.cpp:78]: No need for new line here

**Review by ssanadhya — CHANGES_REQUESTED** [review]

**Comment by rdefosse** [comment]
RestartIntegrationTests

**Comment by 119Vik** [comment]
RestartLTETests

**Comment by 119Vik** [comment]
RestartIntegrationTests

**Comment by 119Vik** [comment]
RestartIntegrationTests

**Comment by 119Vik** [comment]
RestartLTETests

**Comment by 119Vik** [comment]
RestartIntegrationTests

**Comment by rsarwad** [comment]
> Thanks for exploring all options and finishing this issue.

I should be grateful for suggestion from you, Ulas and Shruti

**Comment by rsarwad** [comment]
> RestartIntegrationTests

I assume rebasing to latest master codebase should take care of restarting Integration tests

**Pull request closed without merging**